### PR TITLE
Feat/kvec generator

### PR DESCRIPTION
**Pull request by kvhuguenin**
Implemented class for generating the k-vectors (reciprocal space vectors) used for the k-space implementation of LODE and more general spherical expansion coefficients.
The details and derivations of the algorithm used to generate the vectors is described in a supplementary document that can be provided if desired. 

Tests are provided to check that the code behaves reasonably under realistic conditions as well as some more artificial tests to catch boundary cases.

### CODE REVIEW

File: tests/test_math.hh [hunk 1]
```diff
@@ -458,6 +458,50 @@ namespace rascal {
...
+      this->cells.push_back(cell);
+
+      // Cells 5 and 6: Two triclinic cells
+      cell << 1., 0.3, 0.2, 0.4, 1., -0.1, 0., 0.3, 1.;
+      this->cells.push_back(a * cell);
+      cell << 0.8, 0.3, 0.2, 0.1, 3., -0.2, 0., 0.3, 0.7;
+      this->cells.push_back(a * cell);
+    }
+
+    // Destructori
```
Comment (Luthaf) [inline tests/test_math.hh:498]:
```suggestion
    // Destructor
```

or event better 

```suggestion
```

File: tests/test_math_kvecgenerator.cc [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+    // For small enough cutoff radii, namely for the intervals
+    // [1,sqrt(2)), [sqrt(2), sqrt(3)), [sqrt(3),2) we can explicitly
+    // list all vector that should be found by the algorithm.
+    // Store these solutions in the correct order for each cutoff
+    Eigen::MatrixXd kvec_sol0(3, 3), kvec_sol1(9, 3), kvec_sol2(13, 3);
+    kvec_sol0 << v001, v010, v100;
+    kvec_sol1 << v001, v01m, v010, v011, v1m0, v10m, v100, v101, v110;
+    kvec_sol2 << v001, v01m, v010, v011, v1mm, v1m0, v1m1, v10m, v100, v101,
+        v11m, v110, v111;
+    Eigen::MatrixXd * solutions[3] = {&kvec_sol0, &kvec_sol1, &kvec_sol2};
```
Comment (Luthaf) [inline tests/test_math_kvecgenerator.cc:79]:
```suggestion
    Eigen::MatrixXd kvec_expected_0(3, 3);
    kvec_expected_0 << v001, v010, v100;
    
    Eigen::MatrixXd kvec_expected_1(9, 3);
    kvec_expected_1 << v001, v01m, v010, v011, v1m0, v10m, v100, v101, v110;
    
    Eigen::MatrixXd kvec_expected_2(13, 3);
    kvec_expected_2 << v001, v01m, v010, v011, v1mm, v1m0, v1m1, v10m, v100, v101,
        v11m, v110, v111;
        
    Eigen::MatrixXd * expected[3] = {&kvec_expected_0, &kvec_expected_1, &kvec_expected_2};
```

File: tests/test_math_kvecgenerator.cc [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+
+    // For various cutoff radii close to the thresholds after which
+    // we would get further vectors, we compare the obtained vectors
+    // to the correct solution. The variable i runs over the three intervals
+    double cutofflims[4] = {1., math::SQRT_TWO, math::SQRT_THREE, 2.};
```
Comment (Luthaf) [inline tests/test_math_kvecgenerator.cc:84]: You should use underscores to separate words in your variables names, it makes it much easier to read the code

```suggestion
    double cutoff_lims[4] = {1., math::SQRT_TWO, math::SQRT_THREE, 2.};
```

File: tests/test_math_kvecgenerator.cc [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+    double cutoffeps = 1e-12;
```
Comment (Luthaf) [inline tests/test_math_kvecgenerator.cc:85]:
```suggestion
    double cutoff_eps = 1e-12;
```

File: tests/test_math_kvecgenerator.cc [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+    for (int i = 0; i < 3; i++) {
+      math::Kvectors k_vectorsmin(cutofflims[i] + cutoffeps, reciprocalcell,
```
Comment (Luthaf) [inline tests/test_math_kvecgenerator.cc:87]:
```suggestion
      math::Kvectors k_vectors_min(cutoff_lims[i] + cutoff_eps, reciprocal_cell,
```

File: tests/test_math_kvecgenerator.cc [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+                                  true);
+      BOOST_CHECK_EQUAL(k_vectorsmin.get_kvectors(), *solutions[i]);
+      math::Kvectors k_vectorsmax(cutofflims[i + 1] - cutoffeps, reciprocalcell,
+                                  true);
+      BOOST_CHECK_EQUAL(k_vectorsmax.get_kvectors(), *solutions[i]);
+    }
+
+    // Check whether the code also works if we use a different, highly
+    // distorted basis to describe the same lattice
+    Eigen::Matrix3d cell_distortedbasis;
```
Comment (Luthaf) [inline tests/test_math_kvecgenerator.cc:97]:
```suggestion
    Eigen::Matrix3d cell_distorted_basis;
```

File: tests/test_math_kvecgenerator.cc [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+    double cutoff = 6.28;
+    for (auto & cell : this->cells) {
+      math::Kvectors kvectors(cutoff, cell);
+      int nvec = kvectors.get_numvectors();
+      Eigen::MatrixXd kvecs = kvectors.get_kvectors();
+      Eigen::VectorXd kvals = kvectors.get_norms();
+
+      // Make sure that all vector do indeed lie within
+      // the cutoff and that the lengths computed by the kvecgenerator
+      // agree with those values.
```
Comment (Luthaf) [inline tests/test_math_kvecgenerator.cc:137]: This comment is duplicated from the `@brief` above, no need to have it around twice

File: tests/test_math_kvecgenerator.cc [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+    for (auto & cell : this->cells) {
+      math::Kvectors kvectors(cutoff, cell, false);
+      int nvec = kvectors.get_numvectors();
+      Eigen::MatrixXd kvecs = kvectors.get_kvectors();
+      Eigen::VectorXd kvals = kvectors.get_norms();
+
+      /* Define the used rotation matrices. In order to avoid
+       * discrepancies due to rounding errors, we only use rotations
+       * that are a combination of 90 degrees rotations, since these
+       * rotation matrices only contain 1, 0, and -1 as components.
```
Comment (Luthaf) [inline tests/test_math_kvecgenerator.cc:240]: It might be more interesting to use arbitrary rotation, and check for approximate equality using epsilon ~1e-12. Such test would be able to catch more issues.

File: src/rascal/math/kvec_generator.hh [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+       * These are the quantities that will be used by LODE and the
+       * reciprocal space implementation of SOAP. They are:
+       * - the total number numvectors of vectors within the cutoff,
+       * - the numvectors times 3 Eigen matrix containing those vectors,
+       * - the Eigen vector of size numvectors containing the norms
+       *   of the found vectors.
+       */
+      size_t numvectors = 0;
+      Matrix_t kvecs{};
+      Vector_t kvecnorms{};
```
Comment (Luthaf) [inline src/rascal/math/kvec_generator.hh:81]: Instead of duplicating the information on the number of vectors, you could remove the `numvectors` member, and use `kvecsnorms.size()` to implement `get_numvectors()`.

File: src/rascal/math/kvec_generator.hh [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+      // Eigen::Matrix_Xi kvecindices{};
```
Comment (Luthaf) [inline src/rascal/math/kvec_generator.hh:82]: leftover code?

File: src/rascal/math/kvec_generator.hh [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+
+      // Auxiliary variables for developers/testing:
```
Comment (Luthaf) [inline src/rascal/math/kvec_generator.hh:84]: Can you add a comment for each one of these indicating what they contain/represent?

File: src/rascal/math/kvec_generator.hh [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+        // Run the precompute function described below
+        this->precompute();
+      }  // end constructor
+
+      /**
+       * Using the provided parameters in the constructor, find
+       * all the k-vectors within the cutoff radius and store them
+       * as well as derived quantities that can be returned to the user.
+       */
+      void precompute();
```
Comment (Luthaf) [inline src/rascal/math/kvec_generator.hh:149]: this function is never supposed to be called by the user right? Then it should be at least made private, or even better removed and only be part of the constructor.

File: src/rascal/math/kvec_generator.hh [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+       *     If set to true, it indicates that the cell provided
+       *     in the variable basisvectors is already the reciprocal cell.
+       *     Thus, the formula above for converting the a_i's to the b_i's
+       *     is not used, and the provided cell is used directly.
+       * @param need_origin false by default.
+       *     If set to true, the list of found k-vectors will also
+       *     contain the origin k=(0,0,0), which will then be stored
+       *     as the very first element in the matrix of found vectors.
+       */
+      explicit Kvectors(double cutoff, Eigen::Matrix3d basisvectors,
```
Comment (Luthaf) [inline src/rascal/math/kvec_generator.hh:121]: you don't need an `explicit` here

File: src/rascal/math/kvec_generator.hh [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+                        bool is_reciprocal_cell = false,
+                        bool need_origin = false) {
+        // Set internal variables apart from cell
+        this->kcutoff = cutoff;
+        this->include_origin = need_origin;
+
+        /* Store provided basis vectors into the internal variable basisvecs.
```
Comment (Luthaf) [inline src/rascal/math/kvec_generator.hh:128]: the constructor implementation should be moved to the .cc file, to improve compile time

File: src/rascal/math/kvec_generator.hh [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+       * all k-vectors within cutoff radius without double counting,
+       * where row(i) = i-th vector. If the origin k=(0,0,0) is included,
+       * it is always stored in row(0).
+       */
+      Matrix_t get_kvectors() const { return this->kvecs; }
+
+      /** Returns the Eigen vector of length this->numvectors containing
+       * the norm of all k-vectors within the cutoff without double counting.
+       */
+      Vector_t get_norms() const { return this->kvecnorms; }
```
Comment (Luthaf) [inline src/rascal/math/kvec_generator.hh:183]: I'm not familiar with Eigen enough, but this code looks like it is making a copy. Could we write these two functions to return const references instead?

File: src/rascal/math/kvec_generator.hh [hunk 16]
```diff
@@ -0,0 +1,229 @@
...
+
+      // FUNCTIONS FOR DEVELOPERS / TESTING
```
Comment (Luthaf) [inline src/rascal/math/kvec_generator.hh:185]: If these are not going to be used by the rest of the code and were only useful when developing this class, they should be removed to decrease the amount of code to maintain.

File: src/rascal/math/kvec_generator.hh [hunk 17]
```diff
@@ -0,0 +1,229 @@
...
+        return 2. * M_PI / 3. * kcut * kcut * kcut / sqrt(this->detM);
+      }
+
+      /**
+       * Return an auxilary factor used for tests.
+       * The estimate above obtained from get_numvectors_estimate() clearly
+       * cannot be exactly equal to the actual number of obtained k-vectors.
+       * There however exist rigorous bounds for the deviation between the
+       * estimate and the true value, which we check in the tests.
+       * The details can be found in /tests/test_math_kvec_generators.cc.
```
Comment (Luthaf) [inline src/rascal/math/kvec_generator.hh:206]: if this is used in tests, this functions should be moved there.

File: src/rascal/math/kvec_generator.hh (hunk 15 above)
Comment (agoscinski) [reply src/rascal/math/kvec_generator.hh:183]: As far as I remember one can do this as in
https://github.com/cosmo-epfl/librascal/blob/master/src/rascal/representations/calculator_spherical_expansion.hh#L1368
https://github.com/cosmo-epfl/librascal/blob/master/src/rascal/representations/calculator_spherical_expansion.hh#L1374

using definitions
https://github.com/cosmo-epfl/librascal/blob/master/src/rascal/representations/calculator_spherical_expansion.hh#L222

File: tests/test_math_kvecgenerator.cc (hunk 8 above)
Comment (agoscinski) [reply tests/test_math_kvecgenerator.cc:240]: I haven't seen any orthogonal matrix generator from Eigen, I think you could do

srand to set a fixed seed (as test hyperparameter, might need to include stdlib.h)
(according to eigen mail list https://listengine.tuxfamily.org/lists.tuxfamily.org/eigen/2010/07/msg00403.html)
and use Eigen's for each rotation rotation.setRandom() to make it random and use QR decomposition to get an orthogonal matrix as in
https://eigen.tuxfamily.org/dox/classEigen_1_1HouseholderQR.html#a3e8e56769bbaaed0616ad98c4ff99c7b

File: src/rascal/math/kvec_generator.hh (hunk 16 above)
Comment (kvhuguenin) [reply src/rascal/math/kvec_generator.hh:185]: The norms will actually come in very handy during LODE, so this function has to stay, unless you would propose to recompute all the norms every time. Personally, I had the impression that recomputing the norm for every single vector would be a waste of time vs. the added cost of storing these separately, since the norms are computed during the generation of the vectors anyway.

File: tests/test_math_kvecgenerator.cc (hunk 8 above)
Comment (kvhuguenin) [reply tests/test_math_kvecgenerator.cc:245]: Ok, so this test is a little more tricky than that. Before commit #9c310f3 , what @Luthaf suggests is what I had done as well , namely by explicitly generating some rotation matrices. However, if you want to perform the rotation test on realistic cells, you get into serious rounding error issues.
Most importantly, if you apply a rotation matrix, the number of k-vectors that are found within the cutoff is often not the same any more due to rounding errors. In that case, it makes no sense to check the equality of each pair of vectors, since there will be many vectors that only exist in the rotated cell or vice versa, and checking whether one vector was "left out" would also be a huge pain to make sure you don't fail due to edge cases.

There are some possible solutions to this problem. But before we call it a "problem" too much, I would like to remind you that we typically choose the cutoff to be so large that we don't really care about what happens at the boundary. So if some vector that have length "equal to" kcut aren't found due to these rounding errors, it doesn't really matter for applications.

This is why in commit #9c310f3, I have decided to change the rotation test to something numerically stabler, by choosing rotation matrices that only contain +1, -1 and 0's and thus preserve floating point arithmetic. (this is why in the scaling test, I only use factors of 0.5 and 2, by the way).
With the original rotation test with arbitrary matrices, I had a hand picked set of cells that worked out fine numerically, but I thought that this is not a good long term solution, if the failure of the tests depends too much on the floating point error you get from the specific implementation details (and I had the rotation test fail on me before because I changed something little).

Comment (kvhuguenin) [reply tests/test_math_kvecgenerator.cc:245]: TLDR: Right now, we have a rotation test that is very stringent, i.e. it tests for essentially perfect equality between the results before and after rotation. Such tests are numerically challenging unless you use rotation matrices that do not affect floating point errors, which limits us to compositions of 90 degree rotations in xyz directions.

I did have some more alternatives in mind when I made these changes, which would come in the form of some additional tests for arbitrary rotations. If you want, I could add those. But given that the test file is already longer than both the kvec_generator.cc and .hh files (potentially longer than both of them combined once I move some of the functionality of the .hh file into the tests as suggested in this review), I'm not sure whether this would be more beneficial relative to the amount of extra code to maintain, since we have already 5 tests for a simple class that performs exactly one job.

Comment (agoscinski) [reply tests/test_math_kvecgenerator.cc:245]: Does this problem occur for points exactly on the cutoff? Shouldn't this be fixable by adding a bit of numerical jitter to the cutoff? You use a cutoff which is a multiple of the cell parameters, only in such a case I can see that the double precision error of 1e-16 could change the number of k-vectors. What do you think

Comment (kvhuguenin) [reply tests/test_math_kvecgenerator.cc:245]: It occurs for points close to the cutoff, yes. Back when I still had that in the test, I mostly used a cutoff that was not an integer multiple of the cell parameters, namely 6.28, and I still had these issues.
In principle, it wouldn't be difficult to add jitters to the cutoff until all tests pass, but that was exactly the kind of "fine-tuning" I was trying to avoid by rewriting the tests in the current form, since the tests suddenly started to fail when I changed something else in the code, although I don't recall which cell caused that.

File: src/rascal/math/kvec_generator.hh [hunk 18]
```diff
@@ -0,0 +1,229 @@
...
+       * Return the expected number of obtained vectors based on
+       * continuous approximation, which is simply the volume of a ball
+       * of radius kcut divided by the reciprocal space unit cell.
+       * In practice, we have to divide by 2 since we are only generating
+       * half of the k-vectors, which is why we use 2pi/3 rather than 4pi/3.
+       * The volume of the unit cell in reciprocal space is sqrt(detM).
+       */
+      double get_numvectors_estimate() const {
+        const double kcut = this->kcutoff;
+        return 2. * M_PI / 3. * kcut * kcut * kcut / sqrt(this->detM);
```
Comment (agoscinski) [inline src/rascal/math/kvec_generator.hh:197]: I don't know why, but we have defined our own pi constant in math/utils.hh I would just use it here for consistency within the project

File: src/rascal/math/kvec_generator.hh [hunk 19]
```diff
@@ -0,0 +1,229 @@
...
+        this->include_origin = need_origin;
+
+        /* Store provided basis vectors into the internal variable basisvecs.
+         * We need to distinguish whether the provided vectors are already
+         * reciprocal cell vectors or need to be converted first
+         */
+        if (not is_reciprocal_cell) {  // real space cell is given
+          // Create reciprocal cell from real space cell
+          Eigen::Matrix3d tcell = basisvectors.transpose();
+          this->basisvecs = 2.0 * M_PI * tcell.inverse();
```
Comment (agoscinski) [inline src/rascal/math/kvec_generator.hh:135]: I don't know why, but we have defined our own pi constant in math/utils.hh I would just use it here for consistency within the project

File: tests/test_math_kvecgenerator.cc (hunk 8 above)
Comment (agoscinski) [reply tests/test_math_kvecgenerator.cc:245]: But don't you find it weird: One rotation can only introduce an error in the range 1e-16 (maybe 1e-15), so the k vector count can only change when a k vector is exactly on the cutoff. The probability that this happens is for a random cell should be super low. The basis vector are multiplied by pi so the basis vectors are not a multiple of your cutoffs. I don't understand how the count can change

Comment (kvhuguenin) [reply tests/test_math_kvecgenerator.cc:245]: Yes, I agree. That's why I had printed all the vectors to see whether that was really the issue, and at least then, it was. I probably could have just shifted the cutoff by 1e-10 or so to get rid of that issue. The way I was generating the rotation matrices had huge potential for rounding errors, and combined with the loop over n3 that would accumulate those, I believe that errors on the order of 1e-14 might be realistic actually, but not much more.

**Comment by kvhuguenin** [comment]
Thank you for all the suggestions. I have now pushed a version that has implemented all these changes. The largest two of which are:
  - all functions in the .hh file used only for tests was moved to the "kvecgen_numvectors_reasonable_test". The class now only contains the constructor + the three functions that will be called by LODE in its current implementation. It really is true that the class looks much cleaner now, so thank you @Luthaf for the suggestion!
  - since we had some discussions regarding the rotation test, I have now made a new test using arbitrary rotation matrices by the method suggested by @agoscinski . When I ran the tests, I found that the sup/max/infinity-norm between the original and rotated vectors obtained could be as large as 4e-14 with a relatively small cutoff, so this combined with a very numerical error-prone way to implement rotation matrices originally might have led to my "bad luck" of failing rotation tests in the past.

**Comment by kvhuguenin** [comment]
Right now, for some reason I don't yet understand, the tests compiled with the clang compiler (that's what the clang-n tests are, right?) produce errors which seem to be due to the fact that the "rotation" matrices that are produced by the algorithm are not actual rotation matrices, i.e. they do not satisfy R^TR = RR^T=1. On my machine, the test runs fine. Is anybody familiar with this isse?

**Comment by agoscinski** [comment]
I am looking into this now

File: tests/test_math_kvecgenerator.cc [hunk 20]
```diff
@@ -0,0 +1,339 @@
...
+    cell0 << 2., 0., 0., 0., 1., 0., 0., 0., 3.;
+    cells.push_back(cell0);
+
+    // Generate rotation matrices with fixed seed
+    srand(4653056);
+    size_t num_random = 100;
+    std::vector<Eigen::Matrix3d> rotation_matrices{};
+    for (size_t index = 0; index < num_random; index++) {
+      auto random_matrix = Eigen::Matrix3d::Random();
+      const auto rotation_matrix_0 =
```
Comment (agoscinski) [inline tests/test_math_kvecgenerator.cc:295]:
```suggestion
      Eigen::Matrix3d rotation_matrix_0 =
```
I think auto makes the type `const Eigen::HouseholderSequence<Eigen::Matrix<double, 3, 3, 0, 3, 3>, Eigen::Matrix<double, 3, 1, 0, 3, 1>, 1>`
I am not sure what then in the multiplication later happens, but it is probably not a matrix multiplication

Because auto is interpreted differently interpreted from interpret compiler, it can work on your machine probably using a gnu compiler.

**Comment by kvhuguenin** [comment]
Oh nice! Thank you for the quick fix and lesseon @agoscinski ! I thought that using auto more might increase readability, but I should be more careful in the future then.
Anyway, since this is the first time my code gets reviewed by somebody who actually knows how to do this, let me know if there is anything else that can be improved, both for this code specifically and also in terms of general programming practices @Luthaf and @agoscinski .

File: tests/test_math_kvecgenerator.cc (hunk 20 above)
Comment (Luthaf) [reply tests/test_math_kvecgenerator.cc]: > Because auto is interpreted differently from different compiler, it can work on your machine using a gnu compiler while failing with lang

That's false. `auto` deduction rules are defined by the standard (they are the same rules applied during template substitution) and the same over different compilers (or this is a compiler bug). What might be happening here is that Eigen is doing strange things with templates, which auto take as they are; and using an explicit type forces a cast & evaluate the Eigen template. Overall, that's an Eigen issue more than a compiler/auto issue.

File: src/rascal/math/kvec_generator.hh (hunk 16 above)
Comment (Luthaf) [reply src/rascal/math/kvec_generator.hh]: I was talking about the functions below, in `// FUNCTIONS FOR DEVELOPERS / TESTING`. I agree that the norm makes sense!

File: src/rascal/math/kvec_generator.hh [hunk 21]
```diff
@@ -0,0 +1,137 @@
...
+      Matrix_t get_kvectors() const {
+        return Eigen::Ref<const Matrix_t>(this->kvectors);
+      }
+
+      /** Returns const reference to the Eigen vector containing the norm
+       * of all k-vectors within the cutoff without double counting.
+       */
+      Vector_t get_kvector_norms() const {
+        return Eigen::Ref<const Vector_t>(this->kvector_norms);
+      }
```
Comment (Luthaf) [inline src/rascal/math/kvec_generator.hh:132]: You need to also change the return type of these functions. Here you are creating a const reference, which is then converted to a `Matrix_t`/`Vector_t` value, making a copy.

File: tests/test_math_kvecgenerator.cc (hunk 7 above)
Comment (Luthaf) [reply tests/test_math_kvecgenerator.cc:145]: Still the case 😃!

**Review by Luthaf — COMMENTED** [review]
I have two small nitpicks; and then this should be good to go! 

As @agoscinski mentionned, we will have to either rebase this branch on top of master & cleanup commits, or squash-merge it. Let me know if you want to try to rebase this together.

File: src/rascal/math/kvec_generator.hh [hunk 22]
```diff
@@ -120,14 +120,14 @@ namespace rascal {
        * row(i) = i-th vector. If the origin k=(0,0,0) is included, it is
        * always stored in row(0).
        */
-      Matrix_t get_kvectors() const {
+      Eigen::Ref<const Matrix_t> get_kvectors() {
```
Comment (Luthaf) [inline src/rascal/math/kvec_generator.hh:123]: the function can still be declared as const! Any specific reason why you removed the qualifier?

File: src/rascal/math/kvec_generator.hh [hunk 23]
```diff
@@ -120,14 +120,14 @@ namespace rascal {
...
         return Eigen::Ref<const Matrix_t>(this->kvectors);
       }
 
       /** Returns const reference to the Eigen vector containing the norm
        * of all k-vectors within the cutoff without double counting.
        */
-      Vector_t get_kvector_norms() const {
+      Eigen::Ref<const Vector_t> get_kvector_norms() {
```
Comment (Luthaf) [inline src/rascal/math/kvec_generator.hh:130]: Ideally, the function should be marked as const if it returns a const reference.

```suggestion
      Eigen::Ref<const Vector_t> get_kvector_norms() const {
```
